framework/packages/abstract-adapter/src/endpoints.rs

```diff
@@ -3,35 +3,58 @@ mod ibc_callback;
 mod instantiate;
 mod module_ibc;
 mod query;
-mod receive;
 mod reply;
 mod sudo;
 
 #[macro_export]
 macro_rules! export_endpoints {
     ($api_const:expr, $api_type:ty) => {
-        /// Instantiate entrypoint
+        $crate::__endpoints_without_custom__!($api_const, $api_type);
+        /// Execute entrypoint
         #[::cosmwasm_std::entry_point]
-        pub fn instantiate(
+        pub fn execute(
             deps: ::cosmwasm_std::DepsMut,
             env: ::cosmwasm_std::Env,
             info: ::cosmwasm_std::MessageInfo,
-            msg: <$api_type as $crate::sdk::base::InstantiateEndpoint>::InstantiateMsg,
+            msg: <$api_type as $crate::sdk::base::ExecuteEndpoint>::ExecuteMsg,
         ) -> Result<::cosmwasm_std::Response, <$api_type as $crate::sdk::base::Handler>::Error> {
-            use $crate::sdk::base::InstantiateEndpoint;
-            $api_const.instantiate(deps, env, info, msg)
+            use $crate::sdk::base::ExecuteEndpoint;
+            $api_const.execute(deps, env, info, msg)
         }
-
+    };
+    ($api_const:expr, $api_type:ty, $custom_exec:ty) => {
+        $crate::__endpoints_without_custom__!($api_const, $api_type);
         /// Execute entrypoint
         #[::cosmwasm_std::entry_point]
         pub fn execute(
             deps: ::cosmwasm_std::DepsMut,
             env: ::cosmwasm_std::Env,
             info: ::cosmwasm_std::MessageInfo,
-            msg: <$api_type as $crate::sdk::base::ExecuteEndpoint>::ExecuteMsg,
+            msg: $custom_exec,
         ) -> Result<::cosmwasm_std::Response, <$api_type as $crate::sdk::base::Handler>::Error> {
-            use $crate::sdk::base::ExecuteEndpoint;
-            $api_const.execute(deps, env, info, msg)
+            use $crate::sdk::base::{CustomExecuteHandler, ExecuteEndpoint};
+            match CustomExecuteHandler::try_into_base(msg) {
```

Comment:
> What if we assumed that the base type is present and create the match statement that way?
> 
> We could create a macro that adds the base messages to the custom message.
> 
> So basically we'd do the match in this code
> 
> ```rust
> match msg {
>     $custem_exec::Base(...) => 
>     $custem_exec::Module(...) => 
>     ...
> }
> ```
> 
> If the user doesn't add it then it will err.

I don't see a benefit of doing so, except that we check base variants existing on the module(it can get pretty confusing with ibc variants if we decide to feature flag it).
I like that currently when you create custom message you can see that you covered base variants inside `try_into_base` without doing any macro expansion, and you can set non-custom variants to `unreachable!()` inside `custom_execute`, without hesitation

Reply:
I'm a bit worried about the boilerplate of it, which is why I brought it up. We also encode calls in our SDK with the `ExecuteMsg::Module` variant. If a developers mis-names his/her variant then that could cause issues for consumers.

Reply:
> I'm a bit worried about the boilerplate of it, which is why I brought it up. We also encode calls in our SDK with the `ExecuteMsg::Module` variant. If a developers mis-names his/her variant then that could cause issues for consumers.

Good point! Adding verification of a structure inside schema generation. This will catch miss-names before it gets published